### ISSUE TITLE
Enhanced view of uncommitted changes in Browse Repository

### ISSUE THREAD

**Issue by gerhardol** [gitextensions/gitextensions#4031]
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
It is possible to see staged/working directory changes in the Browse Repository, but the behavior can be improved:
 * Enable the functionality by default: https://git-extensions-documentation.readthedocs.io/en/latest/settings.html#git-extensions-performance-working-dir-changes I added #4033 for this, the possible performance issue can be discussed there too
 * Improve the information displayed (what this feature is about)

Future
 * Move the Commit to modeless as in #564. (a few related details here). 
 * Improve caching and performance

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

**What is the expected behavior?**

 * Diff tab shows changed files and the difference to latest commit. Should show diff to any commit.
 * Clean up for the menu options for virtual commits (Reset, Blame etc)

I have moved some related suggestions to #564. A few of them could be seen as improvements in this feature too.

* Stage/Unstage for the virtual commits
* See number of changed (and new, removed, renamed?) files as in the Commit button

Considered but dropped suggestions:

 * Improve information  for "Commit index" and "Current unstaged changes" in the tree view:
   - See preliminary ("last") commit message
  -> While this could be interesting, it will only be a subset of the Diff tab anyway.

 * See Commit tab, for complete commit message (and parent hash)
  -> Just minor of the info is relevant, quite simple to just select the current HEAD anyway.

 * File tree tab is currently hidden for unstaged. Should show similar info as for a commit (staged the index tree, unstaged just file explorer?. 
  -> There seem to be no obvious way to get the file tree for staged, 'ls-tree' has no --cached option. For unstaged a file explorer could be implemented though.

**Environment you encounter the issue:**
- GitExtensions version: 2.50.02
- GIT version: 2.14.1
- OS version: Win10

**Did this work in previous version of GitExtensions (which)?**
Same in all GitExt

**Comment by jbialobr** [comment]
The working dir changes can be seen from the Browse window, but there is no ability to perform commit related actions. See https://git-extensions-documentation.readthedocs.io/en/latest/settings.html#git-extensions-performance-working-dir-changes

**Comment by gerhardol** [comment]
Great - how did I miss that setting? (Should be on by default - what is slowing down?).

This feature request is then about how this should work - mostly all already marked with "Enhancement". I can rephrase the description.

**Comment by jbialobr** [comment]
>what is slowing down?

I don't know, never have experience a slow down. You should ask the author.

>I can rephrase the description.

Please do that, as it is already very long. Could you also attach sketched pictures of existing windows showing us new buttons, context menus and other changes you would like to do. Perhaps we will create new subissues to discus each proposed change in separation (I predict long discussions here, see #564 it may be helpful).

**Comment by gerhardol** [comment]
Rephrased the description, moved some parts to other issues. For instance editing unstaged commit message could be "implemented in the context of this issue" but at least pull requests should be separate. But most of the basic functionality I wanted is already implemented!

**Comment by RussKie** [comment]
> I don't know, never have experience a slow down. You should ask the
author.

This setting was added 7 years ago by Henk
https://github.com/gitextensions/gitextensions/commit/3d84909f411825fc5f11e9cccb30db3381badbab

**Comment by gerhardol** [comment]
Commit comment:
> This new feature is by default disabled. The revisions are always shown because checking if there are uncommitted changes is just to slow when working in a large repository.
@spdr870 Anything different from GetAllChangedFiles() (that I assume is what slowing down) compared to what is done for the Commit count button?
What is the limiting factor, number of checkout files? What is large? I assume submodules should be included.

@RussKie Thanks for finding this. I am lost in the code (of course, randomly looking at a problem will get one lost).

I believe #4033 should result in default being changed for new installations, similar to other clients (but GitExt can do this better).
Works OK for me (but what is large?)

**Comment by gerhardol** [comment]
I have implemented a rough version of the "Diff" changes. There is currently a lot of special checks for artificial commits, a lot of that handling can be replaced with special diff options. However, some special handling should be added for the menu items (all current options are not relevant like "diff to local").

I updated the description some.

I currently plan to clean up the handling and submit three merge requests:
 * Clean up of artificial commit (will basically enable Diff).
 * Add menu options for stage/unstage
 * Enable artificial by default #4033

**Comment by gerhardol** [comment]
https://github.com/gerhardol/gitextensions/tree/feature/n4031-artificial-commit-diff

A branch with changes I would like to do to improve this handling
I will This does not include enabling the functionality by default (that is #4033) but a few steps to avoid the popup commit as in #564 

The handling for artificial commits was simplified for diff, but code was added to handle artificial commits properly in other situations (there were a lot functions that does not work well with artificial commits already). I also added some new functionality, some inspired from Commit menu options.

I will submit some separate patches to start with. As the changes have evolved together over time and I have rebased and squashed continuously, the changes has not been tested in this order.
(I may retire this branch and start a new if I have not made pull requests yet for the changes).

**Comment by gerhardol** [comment]
First set of PRs submitted. Awaiting primarily #4079 before continuing, will probably close and reopen #4081 due to #4079

The most important commit is https://github.com/gerhardol/gitextensions/commit/7234d28588cad4b22f3dbc6ad58d8542449904c7

**Comment by jbialobr** [comment]
Do you think that this empty place (for artificial revisions) could be used for commit message and commit button?

![gitext](https://user-images.githubusercontent.com/1003909/31877747-eeb7c188-b7d7-11e7-8fee-dd44fcf1ba5a.png)

https://git-extensions-documentation.readthedocs.io/en/latest/settings.html#detailed-browse-repository-window-show-revision-details-next-to-the-revision-list

**Comment by gerhardol** [comment]
@jbialobr Yes, that would be my preferred place. There should be some auto sizing though.
If the user display Commit as a tab, that area should probably be used anyway.
If not using a tab for commit, the area is just empty right now. This is not something I have looked into yet.

I have changed the strategy for merging the changes slightly and submit isolated changes first, the bigger commits later.
#4084 is essential as a cleanup but a little hard to review must be tested.
The core change for this issue has been initially reviewed by @RussKie 
I will need to make some test cases there. Not so much for the current changes as the test cases will be based on examples I use but for regressions. I can see why maintainers may want the test cases before and not trust a promise...

It may be easier to review #4084 and commit 7234d28 together though.
My personal way forward is to merge (or reject) the other pull requests (a few more is coming) and a review of #4084. I can then merge in master to #4084, make a PR with test cases to 7234d28 that can be reviewed and tested.

**Comment by gerhardol** [comment]
Suggestion how to implement "Commit in Browse":

Artificial commits should always be visible, also if no files are staged/unstaged.
Drag&drop enabled (?)

Commit window is in upper right corner, the CommitInput control uses the same panel as commit info but is activated also if the Commit tab is used. In addition, the width of the CommitInfo vs CommitInput control is remembered separately. There should maybe be a minimal height too for CommitInput.
Collapsing/Expanding works fine, adding and removing tabs is a bigger issue.

The CommitInput control is similar to the Commit popup but they can probably not be directly shared. Due to size either a vertical or horizontal (likely) toolbar should be used. The options may have text cut off and converted to tool tip and/or number be limited due to size. There must be button for Commit, Commit&Push, Append, Commit history, Commit template. Branch skipped, options too. Reset likely in menu.
If control is visible in Unstaged, option to stage all too, similar for Staged.

As either staged or unstaged is only seen at once, the visibility of CommitInput may be limited to Staged only. The CommitInfo control should probably disable the Commit button, you do not see what files to be committed otherwise. (Some better ideas please, to avoid switching between the commits.)

#4096 may be a temporary, the CommitInfo is not useful as is.

**Comment by gerhardol** [comment]
Current status, many PRs from this issue, all are not merged yet

Merged, slightly related only but handled in this issue
#4075 
#4085 

Merged general cleanup of icons, internal names
#4076 
#4079 
#4094 
#4105 
#4107 (review comment on https://github.com/gitextensions/gitextensions/commit/7234d28588cad4b22f3dbc6ad58d8542449904c7)

Merged, submodule/artificial cleanup
#4083 
#4088 
#4089 #4092 #4106 (merged on the condition that #4112 is cleaning up the code)
 
Open. Some kind of priority order, not strict
#4087 
#4112 (based on #4087) (UnitTests remains)
#4084 The RevisionGrid changes are more important than the other changes
#4096 
#4086 
#4080 (see also 4. below)

1. Artificial diff to any commit, primarily https://github.com/gitextensions/gitextensions/commit/7234d28588cad4b22f3dbc6ad58d8542449904c7
Dependent on #4112 #4096, to some extent #4084
I have rewritten part of the changes after comments from @RussKie and merged a few changes in other PRs. Will need to add UnitTests. Possibly split in a few separate (dependent) PRs.

2. Make change so Artificial commits are always visible (if option is enabled).

3. Refresh DiffList when explicit actions (stage, reset etc) update it. Ignore external changes still.
Before RevisionDiff refactoring, it was possible to use RefreshRevisions() in FormBrowse. 
If Artificial commits are always visible, just refreshing staged/unstaged is sufficient.

4. Find further inconsistent command handling, for instance #4098 

5. Review the menu action Enable/Visible settings, make more consistent.
A program like GE will have more possibilities than allowed in a commercial program, maybe some limitations should be done.

6. Improve update and performance. Proposal in #4069 that involves updating the count for artificial commits.

7. Add icons for changes for artificial commits, so it is easy to see if i.e. submodules that are dirty/need update?

8. Enable artificial commits by default. #4033

9. Commit panel in Browse window. Put it in the panel where Commit info may be shown for normal commits (see #4096) but show it regardless if user have that option enabled or not. The size for the Commit panel should be remembered individually for artificial and normal commits.
The panel info is a subset of the commit text panel in the commit popup. See the previous comment in this issue for more information.

RussKie *Controller refactoring may be involved somewhere as well.

**Comment by RussKie** [comment]
Awesome, thanks. This should make the life easier :)

**Comment by RussKie** [comment]
@gerhardol do you think we can close the current issue for the work which went into 2.51, and start a new issue for the remaining work which will go into the next version?

**Comment by gerhardol** [comment]
#4209 is connected to this issue, can be connected to a new.
I have seen a few other details too (existing, not regressions so far), separate issue planned.
Tag with 2.51 1nd close it.

**Comment by RussKie** [comment]
Awesome job, mate.
Please open a new issue for further work.
